CC Curve support for standard EG's

PythonBlue opened the pull request:
Implemented a crude approach to supporting CC curves by adding the CurveSet resources to the EG's.

The following opcodes should now be supported with this branch:

ampeg_attack_onccN
ampeg_decay_onccN
ampeg_delay_onccN
ampeg_hold_onccN
ampeg_release_onccN
ampeg_sustain_onccN
fileg_attack_onccN
fileg_decay_onccN
fileg_delay_onccN
fileg_hold_onccN
fileg_elease_onccN
fileg_sustain_onccN
pitcheg_attack_onccN
pitcheg_decay_onccN
pitcheg_delay_onccN
pitcheg_hold_onccN
pitcheg_release_onccN
pitcheg_sustain_onccN

Also, apologies for all the constant re-requests: main issue was my failing to realize I copied all of the files I should have for the branch.

PythonBlue committed:
Initial recommit for personal use

PythonBlue committed:
Update Region.cpp

Fix applied to the CC processing: too many CC values were specified to be per-voice originally.

PythonBlue committed:
Merge remote-tracking branch 'upstream/develop' into develop

PythonBlue committed:
Update Region.cpp

Merged changes in the fork and the main repo

PythonBlue committed:
Merge remote-tracking branch 'upstream/develop' into develop

PythonBlue committed:
Update wavpack

PythonBlue committed:
Preparatory commit for another pull request

For this particular pull request, I'm happy to say I managed to pull off curves for CC-based modulation of ADSR envelope times. Fully tested with amp EG's before committing, and theoretically pitch and filter EG's should work the same.

PythonBlue committed:
Remembered to include one crucial file edit for the eg_curvecc's

PythonBlue committed:
Fixes for PR prep

Replaced references to all resources for EG stages with just the curves, and some other fixes for easy merging

PythonBlue committed:
Update Voice.cpp

Forgot to remove references in my full branch

PythonBlue committed:
Hopefully-final fixes before proper PR

Forgot to implement changes for the Region.cpp file, and also implemented the changes for the Tests for the sake of the whole project building properly.

PythonBlue committed:
Update SynthMessaging.cpp

Fixed SynthMessaging pointing to the new envelope stages.

PythonBlue committed:
Update EGDescription.h

Incorrect default value for ccSustain fixed

paulfd commented:
That's a great feature thanks, and the implem looks good at a glance. I'll expand the existing tests to account for this capability. Is this actually in the spec?

PythonBlue commented:
All of these opcodes are already implemented in ARIA, though your question raises an interesting point that only some of them are listed at all on the sfizz website's opcode page (the one for ampeg attack is missing, for instance?).

redtide commented:
Maybe better here than on Discord:
some opcodes might be missing on sfzformat: the opcode list on sfz.tools uses the "same" data as sfzformat (a copy updated time to time), so if an opcode is missing is just not included in the sfzformat database. Other than that might be that some opcodes marked as unsupported are already implemented, just not keep updated in the sfz.tools database.
So if someone find some error please report in the appropriated website issue tracker, [sfzformat](https://github.com/sfzformat/sfzformat.github.io/issues) or [sfztools](https://github.com/sfztools/sfztools.github.io/issues).

paulfd commented on src/sfizz/EGDescription.h:
```diff
@@ -66,13 +68,20 @@ struct EGDescription {
     CCMap<float> ccStart;
-    CCMap<float> ccSustain;
+    CCMap<ModifierCurvePair<float>> ccSustain { ModifierCurvePair<float>{ Default::egPercent, Default::curveCC } };
+    //CCMap<float> ccAttack;
```
These could probably go away.

PythonBlue replied:
If you're referring to what I commented out, fair point. I just kept it in case revertng was desired.

PythonBlue committed:
Possible CurveCC fixes for PR

While testing a problem I had with an earlier build, that turned out to be an SFZ parsing issue, I decided to apply some code changes to further enforce that custom curves will be read.

PythonBlue committed:
Additional CurveCC fixes

Just noticed the SFZ parser wasn't programmed to read some of them in this branch. Fixed with this commit.

paulfd commented:
Hey @PythonBlue, is this branch in good shape? I can probably fix the tests and look it over shortly if so!

PythonBlue committed:
Update EGDescription.h

Last minute commit before possible merging into main repo. Fixed a type specification issue for the new CC variables that would cause tests to fail, and likely issues with negative values as well.

PythonBlue commented:
Thanks for the reminder about this PR. I THINK I just fixed the tests issue, though I guess we'll find out in a few minutes, but either way, it should be better this way anyway because I realized part of the issue was how the CC Sustain and Start vaues were made.
